### PR TITLE
Fix/issue 2648 Refactor memory module with improved metadata handling and test coverage

### PR DESCRIPTION
**Pull request by johnwlockwood**
This PR introduces several improvements to the memory module:

1. Added new utility functions:

- `create_uuid_mapping()` - Handles temporary UUID to integer mapping
- `unique_old_memory()` - Ensures unique memory entries

2. Enhanced metadata handling:

- Uses deepcopy to prevent modification of original metadata
- Added type hints (dict | None) for metadata parameters

3. Test improvements:

- Completely revamped test_main.py with:

  - Comprehensive test cases for memory operations
  - Detailed test scenarios for both sync and async
  - Better test data organization
  - Improved edge case coverage

4. Code quality:

- Better UUID handling
- Cleaner code organization
- Improved type safety
- Enhanced error handling and logging

Fixes # 2648

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (does not change functionality, e.g. code style improvements, linting)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Please delete options that are not relevant.

- [x] Unit Test

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

## Maintainer Checklist

- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] Made sure Checks passed

### CODE REVIEW

File: mem0/memory/main.py [hunk 1]
```diff
@@ -113,7 +115,7 @@ def add(
         user_id=None,
         agent_id=None,
         run_id=None,
-        metadata=None,
+        metadata: dict | None = None,
```
Comment (johnwlockwood) [inline mem0/memory/main.py:118]: Change this to `Optional[dict]` because this project supports Python >=3.9 and `|` was introduced in Python 3.10

File: mem0/memory/main.py [hunk 2]
```diff
@@ -646,14 +645,17 @@ def history(self, memory_id):
         capture_event("mem0.history", self, {"memory_id": memory_id, "sync_type": "sync"})
         return self.db.get_history(memory_id)
 
-    def _create_memory(self, data, existing_embeddings, metadata=None):
+    def _create_memory(self, data, existing_embeddings, metadata: dict | None = None):
```
Comment (johnwlockwood) [inline mem0/memory/main.py:648]: Change this to `Optional[dict]` because this project supports Python >=3.9 and `|` was introduced in Python 3.10

File: mem0/memory/main.py [hunk 3]
```diff
@@ -667,7 +669,7 @@ def _create_memory(self, data, existing_embeddings, metadata=None):
         capture_event("mem0._create_memory", self, {"memory_id": memory_id, "sync_type": "sync"})
         return memory_id
 
-    def _create_procedural_memory(self, messages, metadata=None, prompt=None):
+    def _create_procedural_memory(self, messages, metadata: dict | None = None, prompt=None):
```
Comment (johnwlockwood) [inline mem0/memory/main.py:672]: Change this to `Optional[dict]` because this project supports Python >=3.9 and `|` was introduced in Python 3.10

File: mem0/memory/main.py [hunk 4]
```diff
@@ -708,16 +711,19 @@ def _create_procedural_memory(self, messages, metadata=None, prompt=None):
 
         return result
 
-    def _update_memory(self, memory_id, data, existing_embeddings, metadata=None):
+    def _update_memory(self, memory_id, data, existing_embeddings, metadata: dict | None = None):
```
Comment (johnwlockwood) [inline mem0/memory/main.py:714]: Change this to `Optional[dict]` because this project supports Python >=3.9 and `|` was introduced in Python 3.10

File: mem0/memory/main.py [hunk 5]
```diff
@@ -850,7 +856,7 @@ async def add(
         user_id=None,
         agent_id=None,
         run_id=None,
-        metadata=None,
+        metadata: dict | None = None,
```
Comment (johnwlockwood) [inline mem0/memory/main.py:859]: Change this to `Optional[dict]` because this project supports Python >=3.9 and `|` was introduced in Python 3.10

File: mem0/memory/main.py [hunk 6]
```diff
@@ -1483,7 +1490,7 @@ async def _create_procedural_memory(self, messages, metadata=None, llm=None, pro
 
         return result
 
-    async def _update_memory(self, memory_id, data, existing_embeddings, metadata=None):
+    async def _update_memory(self, memory_id, data, existing_embeddings, metadata: dict | None = None):
```
Comment (johnwlockwood) [inline mem0/memory/main.py:1493]: Change this to `Optional[dict]` because this project supports Python >=3.9 and `|` was introduced in Python 3.10

File: mem0/memory/main.py [hunk 7]
```diff
@@ -1429,7 +1435,7 @@ async def _create_memory(self, data, existing_embeddings, metadata=None):
         capture_event("mem0._create_memory", self, {"memory_id": memory_id, "sync_type": "async"})
         return memory_id
 
-    async def _create_procedural_memory(self, messages, metadata=None, llm=None, prompt=None):
+    async def _create_procedural_memory(self, messages, metadata: dict | None = None, llm=None, prompt=None):
```
Comment (johnwlockwood) [inline mem0/memory/main.py:1438]: Change this to `Optional[dict]` because this project supports Python >=3.9 and `|` was introduced in Python 3.10

File: mem0/memory/main.py [hunk 8]
```diff
@@ -1404,15 +1407,18 @@ async def history(self, memory_id):
         capture_event("mem0.history", self, {"memory_id": memory_id, "sync_type": "async"})
         return await asyncio.to_thread(self.db.get_history, memory_id)
 
-    async def _create_memory(self, data, existing_embeddings, metadata=None):
+    async def _create_memory(self, data, existing_embeddings, metadata: dict | None = None):
```
Comment (johnwlockwood) [inline mem0/memory/main.py:1410]: Change this to `Optional[dict]` because this project supports Python >=3.9 and `|` was introduced in Python 3.10

File: mem0/memory/main.py [hunk 9]
```diff
@@ -113,7 +115,7 @@ def add(
         user_id=None,
         agent_id=None,
         run_id=None,
-        metadata=None,
+        metadata: Optional[dict] = None,
```
Comment (johnwlockwood) [inline mem0/memory/main.py:118]: A more specific type annotation than `dict` or `Dict` would help developers know the restrictions. 
`metadata` should contain values that are JSON serializable at least.

The underlying vector stores have their own restrictions, so the type guidance here would help developers stay in line and reduce the research or debugging they need to do.

`Supabase.insert` ends up passing metadata to it's `self.collection.upsert` which defines `Metadata` like so:
```
MetadataValues = Union[str, int, float, bool, List[str]]
Metadata = Dict[str, MetadataValues]
```
`CromaDB.insert` also passes metadata to `self.collection.add` which defines `MetaData` with even more restriction:
```
Metadata = Mapping[str, Union[str, int, float, bool]]
```
`Weaviate.insert` handles `metadata` as `WeaviateProperties` which is more permissive. `datetime` isn't JSON serializable by default.:
```
WeaviateField: TypeAlias = Union[
    None,  # null
    str,  # text
    bool,  # boolean
    int,  # int
    float,  # number
    datetime.datetime,  # date
    uuid_package.UUID,  # uuid
    GeoCoordinate,  # geoCoordinates
    Union[PhoneNumber, PhoneNumberType],  # phoneNumber
    Mapping[str, "WeaviateField"],  # object
    Sequence[str],  # text[]
    Sequence[bool],  # boolean[]
    Sequence[int],  # int[]
    Sequence[float],  # number[]
    Sequence[datetime.datetime],  # date[]
    Sequence[uuid_package.UUID],  # uuid[]
    Sequence[Mapping[str, "WeaviateField"]],  # object[]
    # Sequence is covariant while List is not, so we use Sequence here to allow for
    # List[Dict[str, WeaviateField]] to be used interchangeably with List[Dict[str, Any]]
]

WeaviateProperties: TypeAlias = Mapping[str, WeaviateField]
```

`CromaDB`'s `Metadata` is the common denominator, so would be a good choice to model the type on.

**Note**: This doesn't need to be decided in this PR, I'll make an issue.

**Comment by utkarsh240799** [comment]
Hi @johnwlockwood, thank you for opening this PR and for identifying the metadata mutation issue in #2648 — your analysis was spot on.

We've since addressed the core concurrency problem with a minimal targeted fix in #4529, which adds `deepcopy` at the entry points of `_create_memory` and `_create_procedural_memory` (both sync and async) — the same approach you recommended in the issue.

Since that PR covers the safety fix and this one has grown stale with merge conflicts, we're going to close this in favor of #4529. We really appreciate your contribution and the thorough thinking behind it!
